File: apps/passenger-mobile/src/maps/DriverMarkerInterpolator.ts
import type { GeoPoint } from '@/features/maps/types';
import { haversineMeters, lerpBearing, lerpPoint } from './geo';

export type DriverLocationSample = {
  location: GeoPoint;
  bearingDegrees: number | null;
  /** ISO-8601 timestamp from the server (when the GPS fix was taken). */
  recordedAt: string;
};

export type InterpolatorOptions = {
  /** Animation duration for a "normal" update, in ms (MAP_DRIVER_INTERPOLATION_MS). */
  defaultDurationMs: number;
  /** Never animate longer than this, however large the gap between updates. */
  maxDurationMs: number;
  /** A jump implying more speed than this (m/s) is rejected as implausible. */
  maxPlausibleSpeedMetersPerSecond: number;
};

export const DEFAULT_INTERPOLATOR_OPTIONS: InterpolatorOptions = {
  defaultDurationMs: Number(
    process.env.EXPO_PUBLIC_MAP_DRIVER_INTERPOLATION_MS ?? 1_000,
  ),
  maxDurationMs: 5_000,
  maxPlausibleSpeedMetersPerSecond: 55, // ~200 km/h
};

export type IngestOutcome =
  | 'accepted'
  | 'stale_ignored'
  | 'implausible_speed_ignored';

type Keyframe = {
  location: GeoPoint;
  bearingDegrees: number | null;
  recordedAtMs: number;
};

/**
 * Smooths driver marker updates so the marker glides between GPS fixes
 * instead of teleporting on every WebSocket update. Two rejection rules keep
 * the render honest: a sample older than (or equal to) the current target's
 * recordedAt never moves the marker backward without cause, and a sample that
 * implies an impossible speed is rejected outright (kept for the caller to
 * treat as SPOOFING_SUSPECTED via `locationQuality`).
 *
 * Callers ingest server updates via `ingest`, then sample the interpolated
 * render position every frame via `positionAt(nowMs)` — nowMs is always
 * caller-supplied so this class has no hidden clock dependency and is fully
 * deterministic in tests.
 */
export class DriverMarkerInterpolator {
  private from: Keyframe | null = null;
  private to: Keyframe | null = null;
  private animStartMs = 0;
  private animDurationMs = 0;

  constructor(
    private readonly options: InterpolatorOptions = DEFAULT_INTERPOLATOR_OPTIONS,
  ) {}

  ingest(sample: DriverLocationSample, receivedAtMs: number): IngestOutcome {
    const recordedAtMs = Date.parse(sample.recordedAt);

    if (this.to && recordedAtMs <= this.to.recordedAtMs) {
      return 'stale_ignored';
    }

    if (this.to) {
      const elapsedSeconds = (recordedAtMs - this.to.recordedAtMs) / 1_000;
      const distanceMeters = haversineMeters(this.to.location, sample.location);
      const impliedSpeed =
        elapsedSeconds > 0 ? distanceMeters / elapsedSeconds : Infinity;
      if (impliedSpeed > this.options.maxPlausibleSpeedMetersPerSecond) {
        return 'implausible_speed_ignored';
      }
    }

    // The new animation starts from wherever the marker currently sits, so a
    // fast-arriving update never causes a visible snap back to `to`.
    this.from = this.to
      ? { ...this.currentRenderKeyframe(receivedAtMs) }
      : {
          location: sample.location,
          bearingDegrees: sample.bearingDegrees,
          recordedAtMs,
        };

    const previousRecordedAtMs = this.to?.recordedAtMs ?? recordedAtMs;
    const gapMs = recordedAtMs - previousRecordedAtMs;
    this.animDurationMs =
      gapMs > 0
        ? Math.min(gapMs, this.options.maxDurationMs)
        : this.options.defaultDurationMs;
    this.animStartMs = receivedAtMs;
    this.to = {
      location: sample.location,
      bearingDegrees: sample.bearingDegrees,
      recordedAtMs,
    };

    return 'accepted';
  }

  /** The interpolated position to render at wall-clock time `nowMs`. */
  positionAt(nowMs: number): {
    location: GeoPoint;
    bearingDegrees: number | null;
  } {
    if (!this.to) {
      return { location: { latitude: 0, longitude: 0 }, bearingDegrees: null };
    }
    if (!this.from) {
      return {
        location: this.to.location,
        bearingDegrees: this.to.bearingDegrees,
      };
    }

    const t =
      this.animDurationMs <= 0
        ? 1
        : Math.max(
            0,
            Math.min(1, (nowMs - this.animStartMs) / this.animDurationMs),
          );

    return {
      location: lerpPoint(this.from.location, this.to.location, t),
      bearingDegrees: lerpBearing(
        this.from.bearingDegrees,
        this.to.bearingDegrees,
        t,
      ),
    };
  }

  /** True once at least one sample has been accepted. */
  hasPosition(): boolean {
    return this.to !== null;
  }

  private currentRenderKeyframe(nowMs: number): Keyframe {
    const rendered = this.positionAt(nowMs);
    return {
      location: rendered.location,
      bearingDegrees: rendered.bearingDegrees,
      recordedAtMs: this.to?.recordedAtMs ?? nowMs,
    };
  }
}

File: apps/driver-android/core/maps/src/main/kotlin/ru/location/resilienttaxi/driver/core/maps/DriverMarkerInterpolator.kt
package ru.location.resilienttaxi.driver.core.maps

data class DriverLocationSample(
    val location: GeoPoint,
    val bearingDegrees: Double?,
    /** Epoch millis when the GPS fix was taken (server-reported recordedAt). */
    val recordedAtEpochMillis: Long,
)

data class InterpolatorOptions(
    /** Animation duration for a "normal" update, in ms (MAP_DRIVER_INTERPOLATION_MS). */
    val defaultDurationMs: Long = 1_000,
    /** Never animate longer than this, however large the gap between updates. */
    val maxDurationMs: Long = 5_000,
    /** A jump implying more speed than this (m/s) is rejected as implausible. */
    val maxPlausibleSpeedMetersPerSecond: Double = 55.0,
)

enum class IngestOutcome {
    ACCEPTED,
    STALE_IGNORED,
    IMPLAUSIBLE_SPEED_IGNORED,
}

data class RenderedPosition(
    val location: GeoPoint,
    val bearingDegrees: Double?,
)

private data class Keyframe(
    val location: GeoPoint,
    val bearingDegrees: Double?,
    val recordedAtEpochMillis: Long,
)

/**
 * Smooths driver marker updates so the marker glides between GPS fixes
 * instead of teleporting on every location update. A sample older than (or
 * equal to) the current target's recordedAt never moves the marker backward
 * without cause, and a sample implying an impossible speed is rejected
 * outright.
 *
 * Callers ingest server updates via [ingest], then sample the interpolated
 * render position on every render tick via [positionAt] — the wall-clock time
 * is always caller-supplied, so this class has no hidden clock dependency and
 * is fully deterministic in tests.
 */
class DriverMarkerInterpolator(
    private val options: InterpolatorOptions = InterpolatorOptions(),
) {
    private var from: Keyframe? = null
    private var to: Keyframe? = null
    private var animStartMs: Long = 0
    private var animDurationMs: Long = 0

    fun ingest(
        sample: DriverLocationSample,
        receivedAtMs: Long,
    ): IngestOutcome {
        val currentTo = to

        if (currentTo != null && sample.recordedAtEpochMillis <= currentTo.recordedAtEpochMillis) {
            return IngestOutcome.STALE_IGNORED
        }

        if (currentTo != null) {
            val elapsedSeconds = (sample.recordedAtEpochMillis - currentTo.recordedAtEpochMillis) / 1000.0
            val distanceMeters = haversineMeters(currentTo.location, sample.location)
            val impliedSpeed = if (elapsedSeconds > 0) distanceMeters / elapsedSeconds else Double.POSITIVE_INFINITY
            if (impliedSpeed > options.maxPlausibleSpeedMetersPerSecond) {
                return IngestOutcome.IMPLAUSIBLE_SPEED_IGNORED
            }
        }

        // The new animation starts from wherever the marker currently sits, so
        // a fast-arriving update never causes a visible snap back to `to`.
        from =
            if (currentTo != null) {
                val rendered = positionAt(receivedAtMs)
                Keyframe(rendered.location, rendered.bearingDegrees, currentTo.recordedAtEpochMillis)
            } else {
                Keyframe(sample.location, sample.bearingDegrees, sample.recordedAtEpochMillis)
            }

        val previousRecordedAtMs = currentTo?.recordedAtEpochMillis ?: sample.recordedAtEpochMillis
        val gapMs = sample.recordedAtEpochMillis - previousRecordedAtMs
        animDurationMs = if (gapMs > 0) minOf(gapMs, options.maxDurationMs) else options.defaultDurationMs
        animStartMs = receivedAtMs
        to = Keyframe(sample.location, sample.bearingDegrees, sample.recordedAtEpochMillis)

        return IngestOutcome.ACCEPTED
    }

    /** The interpolated position to render at wall-clock time `nowMs`. */
    fun positionAt(nowMs: Long): RenderedPosition {
        val currentTo = to ?: return RenderedPosition(GeoPoint(0.0, 0.0), null)
        val currentFrom = from ?: return RenderedPosition(currentTo.location, currentTo.bearingDegrees)

        val t =
            if (animDurationMs <= 0) {
                1.0
            } else {
                ((nowMs - animStartMs).toDouble() / animDurationMs).coerceIn(0.0, 1.0)
            }

        return RenderedPosition(
            location = lerpPoint(currentFrom.location, currentTo.location, t),
            bearingDegrees = lerpBearing(currentFrom.bearingDegrees, currentTo.bearingDegrees, t),
        )
    }

    /** True once at least one sample has been accepted. */
    fun hasPosition(): Boolean = to != null
}
